module/input/Microphone/src/Microphone.cpp

```diff
@@ -0,0 +1,367 @@
+        write(handles.stdin, input, strlen(input));
+        
+        bytes_read = read(handles.stdout, buffer, sizeof(buffer)-1);
+        assert(bytes_read > 0);
+        buffer[bytes_read] = 0;
+        
+        output = strdup(buffer);
+    }
+    if(handles.stdout) close(handles.stdout);
+    if(handles.stderr) close(handles.stderr);
+    if(handles.stdin) close(handles.stdin);
+
+    return output;
+}
+
+void write_audio_to_file(int fd, char *filename) {
+    FILE *file = fopen(filename, "rb");
+    fseek(file, 0, SEEK_END);
+    int file_size = ftell(file);
+    fseek(file, 0, SEEK_SET);
+    assert(file_size != -1);
+    
+    char *file_buffer = (char*) malloc(file_size);
+    fread(file_buffer, file_size, 1, file);
+    fclose(file);
+    
+    char file_size_buf[64];
+    sprintf(file_size_buf, "%d\n", file_size);
+        
+    write(fd, file_size_buf, strlen(file_size_buf));
+    
+    ssize_t write_res = write(fd, file_buffer, file_size);
+    assert(write_res == file_size);
+
+    free(file_buffer);
+}
+
+
+Microphone::Microphone(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)), config{} {
+    
+    on<Configuration>("Microphone.yaml").then([this](const Configuration& cfg) {
+        // Use configuration here from file Microphone.yaml
+        this->log_level = cfg["log_level"].as<NUClear::LogLevel>();
+    });
+    
+
+    
+    if(!voice2json_transcribe_wav(&handles)) {
+        perror("failed to start voice2json");
+        return;
+    }
+    
+    on<Trigger<MicControlMsg>>().then([this](const MicControlMsg& msg) {
+        switch(msg.type) {
+            case MIC_MSG_ENABLE: 
+                enabled = true;
+            break;
+            case MIC_MSG_DISABLE:
+                enabled = false;
+            break;
+            case MIC_MSG_TEST_AUDIO:
+                char filename[1024];
+                sprintf(filename, "/home/nubots/nuhear/audio/voice_%03d.wav", msg.data);
+            
+                write_audio_to_file(handles.stdin, filename);
+            break;
+        }
+    });
+    
+    //THIS IS FOR TESTING
+    on<Trigger<SpeechIntentMessage>>().then([this](const SpeechIntentMessage &msg) {
+        //TODO needs to be freed...
+        printf("text = %s (%p)\n", msg.text.data(), msg.text.data());
+        printf("intent = %s (%p)\n", msg.intent.data(), msg.intent.data());
+        for(Slot slot : msg.slots) {
+            printf("%s = %s\n", slot.name.data(), slot.value.data());
+        }
+        //need to free the msg...
+        //delete &msg;
+    });
+
+    
+    on<IO>(handles.stdout, IO::READ).then([this] {
+        printf("fds[0].revents & POLLIN\n");
+            
+        //TODO: need to read and append to buffer until we reach a newline...
+        char buffer[0x1000];
+        ssize_t bytes_read = read(handles.stdout, buffer, sizeof(buffer)-1);
+        if(bytes_read > 0) {
+            assert((size_t)bytes_read < sizeof(buffer)); //TODO
```

Comment:
Can this assert ever be triggered? I thought `read` will guarantee not to overfill the buffer.

Reply:
Overzealous checking on my part.  Changed to checking for -1 error condition and logs and error in that case.